elements/pl-hidden-hints/pl-hidden-hints.py

```diff
@@ -0,0 +1,102 @@
+    hints_dict['submission_count'] = 0
+
+    # Parse hints from frontend
+    element = lxml.html.fragment_fromstring(element_html)
+
+    html_hints = []
+
+    # Read hints defined as child tags and load them into html_hints
+    for i, child in enumerate(element):
+        if child.tag == 'pl-hint':
+            pl.check_attribs(child, ['show-after'], [])
+            priority = pl.get_integer_attrib(child, 'show-after')
+            html_hints.append((priority, pl.inner_html(child)))
+        else:
+            raise Exception(f"Child tag type '{child.tag}' not supported by hidden-hints.")
+
+    hints_dict['html_hints'] = html_hints
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
+    # No need to display hints on answer / submission panels
+
+    if data['panel'] == 'question':
+        hints_dict: HintsDict = data['params']['hidden_hints']
+
+        html_hints = hints_dict['html_hints']
+
+        # A little bit of a hack to make it so that all of the backend hints are closed.
+        param_hints = [(-1, hint) for hint in data['params'].get('hints', [])]
+
+        if param_hints and html_hints:
+            raise ValueError('Hints given as both HTML and parameters. Only one can be used.')
+
+        # Sort based on priority
+        hint_list = sorted(html_hints) if html_hints else param_hints
+        hints_to_display = []
+
+        submission_count = hints_dict['submission_count']
+
+        all_correct = all_questions_correct(data)
+
+        # Parse through hint list to display hints.
+        for (idx, (priority, hint)) in enumerate(hint_list):
+            if priority <= submission_count:
+                # Close hints once all questions are correct
+                show_open = not all_correct and priority == submission_count
+                hints_to_display.append({
+                    'hint': hint,
+                    'index': idx + 1,
+                    'is_open': show_open
+                })
+
+        with open('pl-hidden-hints.mustache', 'r') as f:
+            return chevron.render(f, {
+                'hints': hints_to_display,
+                'is_plural': len(hints_to_display) > 1
+            }).strip()
+
+    else:
+        return ''
+
+
+def all_questions_correct(data: pl.QuestionData) -> bool:
+    """Return True if all questions are correct in partial scores and it's nonempty."""
+    partial_scores = data['partial_scores']
+
+    if len(partial_scores) == 0:
+        return False
+
+    return all(part['score'] == 1.0 for part in partial_scores.values())
+
+
+def grade(element_html: str, data: pl.QuestionData) -> None:
+    hints_dict: HintsDict = data['params']['hidden_hints']
+
+    # Only reveal next hint if the previous answer was valid
+    # (i.e. didn't contain any format errors) but incorrect.
+    if (len(data['format_errors']) == 0 and not all_questions_correct(data)):
+        hints_dict['submission_count'] += 1
```

Comment:
I think this could have unexpected behavior if e.g. this element is placed between two graded elements: when `<pl-hidden-hints>` is graded, only the first one would be represented in `partial_scores`.

Separately, I'm a little hesitant to use calls to `grade()` as a proxy for "how many submissions have been made". I think that will generally hold true today, but improved regrading features in the future could result in us calling `grade()` multiple times per submission, even potentially out of order.

Reply:
I think these concerns are valid. When we were working with our own course this was fine, but I think for this version it might be worth adding a "valid submissions so far" counter to the data dict.

Reply:
Got this working by exposing the internal PL submission counter to the data dict!